src/pem_decoder.rs

```diff
@@ -0,0 +1,200 @@
+          // "there is no such thing as a "PKCS#1 format" for elliptic curve (EC) keys"
+
+          // This handles PKCS#8 private keys
+          "PRIVATE KEY" => {
+            match classify_pem(&asn1_content) {
+              Option::Some(Classification::EC) => Ok(PemEncodedKey {
+                content: pem_contents,
+                asn1: asn1_content,
+                pem_type: PemType::ECPrivateKey,
+                encoded_with: PemEncodedWith::PKCS8,
+              }),
+              Option::Some(Classification::RSA) => Ok(PemEncodedKey {
+                content: pem_contents,
+                asn1: asn1_content,
+                pem_type: PemType::RSAPrivateKey,
+                encoded_with: PemEncodedWith::PKCS8,
+              }),
+              _ => return Err(ErrorKind::InvalidKeyFormat)?,
+            }
+          }
+
+          // This handles PKCS#8 public keys
+          "PUBLIC KEY" => {
+            match classify_pem(&asn1_content) {
+              Option::Some(Classification::EC) => Ok(PemEncodedKey {
+                content: pem_contents,
+                asn1: asn1_content,
+                pem_type: PemType::ECPublicKey,
+                encoded_with: PemEncodedWith::PKCS8,
+              }),
+              Option::Some(Classification::RSA) => Ok(PemEncodedKey {
+                content: pem_contents,
+                asn1: asn1_content,
+                pem_type: PemType::RSAPublicKey,
+                encoded_with: PemEncodedWith::PKCS8,
+              }),
+              _ => return Err(ErrorKind::InvalidKeyFormat)?,
+            }
+          }
+
+          // Unknown type
+          _ => return Err(ErrorKind::InvalidKeyFormat)?,
+        }
+      },
+      Err(_) => return Err(ErrorKind::InvalidKeyFormat)?,
+    }
+  }
+
+  /// This will do the initial parsing of a PEM file.
+  /// Supported tagged pems include "RSA PRIVATE KEY", "RSA PUBLIC KEY",
+  /// "PRIVATE KEY", "PUBLIC KEY"
+  /// PEMs with multiple tagged portions are not supported
+  pub fn as_key(&self) -> Result<Key<'_>> {
+    match self.encoded_with {
+      PemEncodedWith::PKCS1 => Ok(Key::Der(self.content.as_slice())),
+      PemEncodedWith::PKCS8 => {
+        match self.pem_type {
+          PemType::RSAPrivateKey => Ok(Key::Der(extract_first_bitstring(&self.asn1)?)),
+          PemType::RSAPublicKey => Ok(Key::Der(extract_first_bitstring(&self.asn1)?)),
+          PemType::ECPrivateKey => Ok(Key::Pkcs8(self.content.as_slice())),
+          PemType::ECPublicKey => Ok(Key::Pkcs8(extract_first_bitstring(&self.asn1)?)),
+        }
+      },
+    }
+  }
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+/// Supported PEM files for EC and RSA Public and Private Keys
+enum PemType {
+  ECPublicKey,
+  ECPrivateKey,
+  RSAPublicKey,
+  RSAPrivateKey,
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+enum PemEncodedWith {
+  PKCS1,
+  PKCS8,
+}
+
+#[derive(Debug)]
+#[derive(PartialEq)]
```

Comment:
You can (and should) put those derive in a single command: `#[derive(Debug, PartialEq)]`